### PR TITLE
WIP: Test suite setup improvement

### PR DESCRIPTION
**Pull request by ikwattro**
Refactored test suite setup
- Adjusted namespace definition in composer.json
- Corrected namespaces in Neo4jClientMock, as \ before fqdcn are not required when using namespaces correctly
- Added a .travis.yml file for using Travis and having a real db to test on
- Added a Import file to import data when the Slim bootstrap is done

I just need to look a bit deeper of how to bootstrap the app correctly for integration tests.

Running `phpunit` is ok.

Would be cool if merged to add `Travis` service webhook

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+language: php
+
+php:
+    - 5.4
+    - 5.5
```
Comment (jeremykendall) [inline .travis.yml:6]: Please add PHP 5.3 and PHP 5.6.

File: composer.json [hunk 2]
```diff
@@ -50,7 +50,7 @@
     },
     "autoload-dev": {
         "psr-4": {
-            "GraphStory\\Tests\\": "tests/"
+            "GraphStory\\GraphKit\\Tests\\": "tests/"
```
Comment (jeremykendall) [inline composer.json]: /me slaps forehead

File: .travis.yml (hunk 1 above)
Comment (ikwattro) [inline .travis.yml:6]: 5.3 is EOL. But ok :)

File: tests/GraphKitTestCase.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace GraphStory\GraphKit\Tests;
+
+use Everyman\Neo4j\Client;
```
Comment (jeremykendall) [inline tests/GraphKitTestCase.php:5]: You can kill this `use` since you're using the fully qualified name in the method. Or use an alias if you like.

**Comment by ikwattro** [comment]
5.3 & 5.6 added to .travis

File: tests/GraphKitTestCase.php (hunk 3 above)
Comment (ikwattro) [inline tests/GraphKitTestCase.php:5]: My bad, I will add a commit using Client in the method, copy-pasted whooooooo

Comment (jeremykendall) [inline tests/GraphKitTestCase.php:5]: > copy-pasted whooooooo

I get caught by that too :-)

**Comment by ikwattro** [comment]
I have added a first test for the UserService.
- One point though. Aren't you handling the case where the username already exist ?

**Comment by jeremykendall** [comment]
> Aren't you handling the case where the username already exist ?

Do you mean this? https://github.com/GraphStory/graph-kit-php/blob/develop/public/index.php#L170-L187

**Comment by ikwattro** [comment]
Mmmmh no. If i do : 

``` php
$service = new UserService();
$user = new User();
$user->id = 123;
$user->firstname = 'xxx';
$user->username = 'alto';
$user->lastname = 'yyy';
$service->save($user);
```

and the user already exist in the database. Then you receive a raw api http response of constraint violation.

I think you shoud try/catch this exception and handle your proper exception for this case.

**Comment by ikwattro** [comment]
Please note that with the last commit, running the `phpunit` command will clear your local database and populate it with a person -> content -> tag schema based on GraphStory demo dataset.

So be aware of it !

This is what the schema looks like in test suite : 

Imgur

**Comment by ikwattro** [comment]
All the methods from the User Service are tested against a test database data. and tests are passing ;-)

**Comment by jeremykendall** [comment]
:+1:

**Comment by jeremykendall** [comment]
> I think you shoud try/catch this exception and handle your proper exception for this case.

Agreed.

**Comment by jeremykendall** [comment]
> Please note that with the last commit, running the phpunit command will clear your local database

Running the unit tests requires a local install of Neo4j?

**Comment by ikwattro** [comment]
Yes

**Comment by ikwattro** [comment]
Testing following/removeFriend and so on, does not make real sense with mocking IMO.

**Comment by ikwattro** [comment]
If you want, we can let PHPUnit do only unit testing and not integration testing.

Maybe integrating Behat+Mink+Zombie.js could be a nice stack for Integration.

Also we can provide a default config file that you should fill in before running the tests with databases connection settings, so if you want you can run the tests against a remote database.

And add a check if the TRAVIS env variable is found, then run against localhost automatically

**Comment by jeremykendall** [comment]
I have zero problem using PHPUnit for some integration testing. It'll also help keep the lib a little more simple to opt against adding Behat+Mink+Zombie.js. We should probably make it very clear which are unit tests and which are integration/functional tests. I prefer using the [`@group` annotation](https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.group) for that purpose.

The app already knows to load files based on environment, as long as `SLIM_MODE` is available as an env variable (see https://github.com/GraphStory/graph-kit-php/blob/develop/public/index.php#L27-L31).

**Comment by jeremykendall** [comment]
Here's an example `.travis.yml` from a project of mine making use of PHPUnit groups: https://github.com/jeremykendall/flaming-archer/blob/develop/.travis.yml#L7

Here's the test it's skipping: https://github.com/jeremykendall/flaming-archer/blob/85f8668a2e45ce792b0ef99d285ba57a52214b10/tests/FA/Tests/Service/FlickrServiceIntegrationTest.php#L13

**Comment by ikwattro** [comment]
Will check for that ;-) 

For the env mode, yeah I saw in the Slim Doc.

**Comment by jeremykendall** [comment]
I prepended WIP (Work In Progress) to the PR name so I don't merge before you're ready. Update the name to remove WIP when you're ready for me to review and merge.

**Comment by ikwattro** [comment]
Yes for sure. BTW I will focus on the UserService and basic unit testing, in order to have a common vision of the test suite ;-)

**Comment by jeremykendall** [comment]
:+1: 

@guillermoandrae, feel free to jump in with any input/discussion on the test suite. I know you were thinking about adding some functional tests for Slim.

File: phpunit.xml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+<phpunit bootstrap="./vendor/autoload.php" colors="true" verbose="true">
```
Comment (guillermoandrae) [inline phpunit.xml]: This will probably change once we get some functional tests involved. We may want to define a few settings in a proper bootstrap file.

Comment (ikwattro) [inline phpunit.xml]: yeah sure. also moving a big part of index.php under the app folder would
ease env management

Christophe Willemsen
Le 31 oct. 2014 04:33, "Guillermo A. Fisher" notifications@github.com a
écrit :

Comment (jeremykendall) [inline phpunit.xml]: > also moving a big part of index.php under the app folder would ease env management

@ikwattro - Huh? I don't understand this comment.

Comment (ikwattro) [inline phpunit.xml]: I mean having an /app dir where you would move the big processing done in the index.php file, and from the index you just require the files from the app_dir.

for example you can have in app/ 

```
- app/
-- config_dev.php
-- config_test.php
-- config_prod.php
-- controllers/
--- controller_dev.php
--- controller_prod.php
```

And from the index you load the files based on the SLIM_MODE

Comment (jeremykendall) [inline phpunit.xml]: Aha. I get it now.

I prefer, in Slim applications, to try and refactor all of the big processing and business logic out of `public/index.php` and into the service layer. Configuration (and dependency injection) should then be wholly responsible for handling environment management.

It's an excellent idea, but I'd like to try and stick with the Slim PHP paradigm as much as possible in this app.

Comment (ikwattro) [inline phpunit.xml]: Cool. I do not have a lot of experience with slim DI, btw I'll look into the doc, should not be worth as I'm far comfortable with the Symfony DIC.

**Comment by jeremykendall** [comment]
@ikwattro - I had to change the `GraphStory\GraphKit\Tests` namespace from plural to singular: `GraphStory\GraphKit\Test` in the develop branch. Having the namespace `Tests` in the `/tests` directory made php-cs-fixer's psr-0 fixer think the namespace should have been `GraphStory\GraphKit\tests` (note the lowercase `tests`). That's happening on my MBP, which might be significant since OSX generally uses a case-insensitive file system.

Updating this PR based on the following diffs _should_ resolve the merge conflict that will otherwise occur when this is finally merged into dev.

``` diff
diff --git a/tests/Mock/Neo4jClientMock.php b/tests/Mock/Neo4jClientMock.php
index bcd9818..ec83143 100644
--- a/tests/Mock/Neo4jClientMock.php
+++ b/tests/Mock/Neo4jClientMock.php
@@ -1,8 +1,8 @@
 <?php

-namespace GraphStory\Tests\Mock;
+namespace GraphStory\GraphKit\Test\Mock;

-use \Everyman\Neo4j\Client;
+use Everyman\Neo4j\Client;

 class Neo4jClientMock extends Client
 {
diff --git a/tests/Unit/Neo4jClientTest.php b/tests/Unit/Neo4jClientTest.php
index 9ee5f90..c4a9dbd 100644
--- a/tests/Unit/Neo4jClientTest.php
+++ b/tests/Unit/Neo4jClientTest.php
@@ -1,9 +1,9 @@
 <?php

-namespace GraphStory\Tests\Unit;
+namespace GraphStory\GraphKit\Test\Unit;

-use \GraphStory\GraphKit\Neo4jClient;
-use \GraphStory\Tests\Mock\Neo4jClientMock;
+use GraphStory\GraphKit\Neo4jClient;
+use GraphStory\GraphKit\Test\Mock\Neo4jClientMock;

 class Neo4jClientTest extends \PHPUnit_Framework_TestCase
 {
@@ -20,6 +20,7 @@ class Neo4jClientTest extends \PHPUnit_Framework_TestCase
     {
         if (!$this->mockClient) {
             $this->mockClient = new Neo4jClientMock();
+
             return $this->mockClient;
         }
     }
```

``` diff
diff --git a/composer.json b/composer.json
index 08d2026..7e5e14a 100644
--- a/composer.json
+++ b/composer.json
@@ -52,7 +52,7 @@
     },
     "autoload-dev": {
         "psr-4": {
-            "GraphStory\\Tests\\": "tests/"
+            "GraphStory\\GraphKit\\Test\\": "tests/"
         }
     }
 }
```

File: phpunit.xml (hunk 4 above)
Comment (jeremykendall) [inline phpunit.xml]: [Slim DI](http://docs.slimframework.com/#DI-Overview) is closer to a Service Locator than textbook Dependency Injection. I've used both [Pimple](http://pimple.sensiolabs.org/) and [Aura.Di](https://github.com/auraphp/Aura.Di) with Slim and have been happy with both.

**Comment by ikwattro** [comment]
Mmmmh I'm not ok. For 2 reasons : 
1. You use `psr-4` in your composer, so it MUST stick to psr-4
2. PSR-0 is deprecated

```
Deprecated - As of 2014-10-21 PSR-0 has been marked as deprecated. PSR-4 is now recommended as an alternative.
```

http://www.php-fig.org/psr/psr-0/

BTW, I do not see the need to run php-cs-fixer against the tests directory

**Comment by jeremykendall** [comment]
> Mmmmh I'm not ok. For 2 reasons :
> 
> You use psr-4 in your composer, so it MUST stick to psr-4

True, however, if you create the correct directory structure and use the correct prefixes in `composer.json`, PSR-0 and PSR-4 are [perfectly compatible](https://philsturgeon.uk/blog/2014/02/phpfig-autoloaders-amendments-and-the-15th-standard), so becomes a non-issue and ...

> PSR-0 is deprecated

Also true, but that's not precisely at issue here. The problem is that php-cs-fixer does not have a psr-4 fixer (yet?). The psr-0 fixer will leave namespaces alone, except in what seems to be an edge-case that we've stumbled upon, but it will change the classname to match the filename if there is a mismatch, which is a good thing.

_However_, all of that said, unit tests will catch misnamed classes and improper namespaces by exploding, so perhaps it's not the big deal I thought it was at first.

If you feel strongly enough that we should go back to `GraphStory\GraphKit\Tests`, then we'll do it _and_ remove the psr-0 fixer in the `.php_cs` file.

> BTW, I do not see the need to run php-cs-fixer against the tests directory

Probably not, but I'm really picky about PSR compliant code, so I usually do.

**Comment by ikwattro** [comment]
PSR-0 and PSR-4 are fully compatible true. Now as I said I never run the fixer on tests directory, but you're the boss.

Keeping a psr-0 fixer is better than not having it while psr-4 is not yet handle AFAIK.

So...

I'll rebase on develop then ;-) It's tests directory, my opinion would not have been the same for the src dir.

**Comment by ikwattro** [comment]
Rebased ;-)

**Comment by jeremykendall** [comment]
:+1:

**Comment by ikwattro** [comment]
@jeremykendall what would be a name you would like for grouping the tests using a real database ?

I thought about `graphdb` ?

**Comment by jeremykendall** [comment]
I like `functional` or `integration`. While those aren't precisely the same thing, I'm comfortable with either.

**Comment by guillermoandrae** [comment]
+1 for `integration`.

**Comment by ikwattro** [comment]
Well `integration` tests checks communication between the different modules of your application while `functional` tests are use case scenarios against a small test of data.

Now whatever we call it, do I group only the tests related to interaction with the database ? And for all other tests I use the client mock ? Am I right ?

**Comment by jeremykendall** [comment]
> Now whatever we call it, do I group only the tests related to interaction with the database ? And for all other tests I use the client mock ? Am I right?

That sounds right to me, @ikwattro. Feel free to name and group as you see fit, and we'll go from there.

"Make it work, refactor!", I like to say :-)

**Comment by ikwattro** [comment]
Well you stay leader hey :) :) :)

**Comment by jeremykendall** [comment]
True, but I don't know everything. That's why I'm so glad you and @guillermoandrae and @michaelmoussa have been bringing so many awesome ideas.

**Comment by michaelmoussa** [comment]
Easy to bring ideas when there's already a good base @jeremykendall :)

**Comment by ikwattro** [comment]
You can now run tests by excluding the `integration` group : 

``` bash
phpunit --exclude-group integration
```

However, regarding unit testing, there is not so much to do. Service classes contains too much responsability, IMO queries should move to a QueryRepository directory and model classes are not OO classes.

I would be pleased if someone could add a running webTestCase bootstrapping the Slim app environment so I can add tests on it.

**Comment by jeremykendall** [comment]
SWEET

File: tests/GKP_IMPORT.cqs [hunk 5]
```diff
@@ -0,0 +1,1189 @@
...
+SET n68.tag = 'sunt';
+MERGE (n69:Tag {neogen_id: '91064f03c78f5a8f604f0a909d457b06d6f8b609' })
+SET n69.tag = 'vel';
+MERGE (n70:Tag {neogen_id: '8f5c4cde829f6b81197553eb93c7697f1ae292b0' })
+SET n70.tag = 'assumenda';
+MATCH (s1:User {neogen_id: '9bf768bcd7e3087618bd72386d87bc224042dd27'}), (e1:Content { neogen_id: '7bd0db8573f7d118ed6f8973f566fdf1165b936c'})
+MERGE (s1)-[edge1:LAST_POST]->(e1)
+;
+MATCH (s2:User {neogen_id: 'c18d6287571873d47fe6650f208702d1dd62a4a2'}), (e2:Content { neogen_id: '27f0f3f74117b38052d63295d98f041a7beea319'})
+MERGE (s2)-[edge2:LAST_POST]->(e2)
```
Comment (jeremykendall) [inline tests/GKP_IMPORT.cqs:149]: This type should be `LASTPOST`.

File: tests/GKP_IMPORT.cqs [hunk 6]
```diff
@@ -0,0 +1,1189 @@
...
+MERGE (s18)-[edge18:LAST_POST]->(e18)
+;
+MATCH (s19:User {neogen_id: '7a518ddb42a5867681b340c5ba3efa500c63b316'}), (e19:Content { neogen_id: '0d21b8c9e468487bb700294e8c83ef3d02a2b15e'})
+MERGE (s19)-[edge19:LAST_POST]->(e19)
+;
+MATCH (s20:User {neogen_id: '7a518ddb42a5867681b340c5ba3efa500c63b316'}), (e20:Content { neogen_id: '67ce21605b87ccf4e4a3745bf47ef6ad2624311b'})
+MERGE (s20)-[edge20:LAST_POST]->(e20)
+;
+MATCH (s21:Content {neogen_id: '7bd0db8573f7d118ed6f8973f566fdf1165b936c'}), (e21:Content { neogen_id: 'a9261ee0919c5a93baedfed2e88361f3d3f29688'})
+MERGE (s21)-[edge21:NEXT]->(e21)
```
Comment (jeremykendall) [inline tests/GKP_IMPORT.cqs:206]: This type should be `NEXTPOST`.

**Comment by jeremykendall** [comment]
> ... regarding unit testing, there is not so much to do. Service classes contains too much responsability, IMO queries should move to a QueryRepository directory and model classes are not OO classes.

True, there's a lot of refactoring that can be done. I want to let TDD guide refactoring as 1) time permits for myself and 2) contributors are willing to refactor.

> I would be pleased if someone could add a running webTestCase bootstrapping the Slim app environment so I can add tests on it.

@guillermoandrae Is this something you'd be interested in?

**Comment by guillermoandrae** [comment]
@jeremykendall  http://www.quickmeme.com/img/8d/8d177ac8acf5d8de0ed4d0e741a354459733a44ce51be37efe457c510bd9be6a.jpg

**Comment by jeremykendall** [comment]
image

**Comment by jeremykendall** [comment]
@ikwattro As soon as you feel like this is good to go, remove the WIP and let's merge it. Then we can move on to adding other tests and refactoring current code and such. I don't want to do too much in a single PR.

**Comment by ikwattro** [comment]
Well I'll remove the wip once there is a webTestCase in place. Currently I can go further with other xxServiceTests as they are in the @integration group

**Comment by jeremykendall** [comment]
:+1: 

@guillermoandrae, I think @ikwattro just called you out ;-)

**Comment by ikwattro** [comment]
:)

**Comment by guillermoandrae** [comment]
lol

@ikwattro + @jeremykendall in which directory do you guys want me to put it? And should I do it in a new PR? New branch?

**Comment by ikwattro** [comment]
A `WebTestCase` file under the `tests/GraphStory\GraphKit\Test` directory sounds good to me.

Having the setUp() function there building the slim app, so we can override it in tests classes extending the WebTestCase if needed.

**Comment by ikwattro** [comment]
You can push to this P.R., no problem

**Comment by jeremykendall** [comment]
make-it-so-captain

**Comment by guillermoandrae** [comment]
@ikwattro what do you think about merging this PR in so I can just create a new PR with all of your goodies in it?

**Comment by guillermoandrae** [comment]
@ikwattro if you're not on, I'm just gonna merge in your changes locally and refer to this pull request when I make my new one.

**Comment by ikwattro** [comment]
Ok for me ;-)

**Comment by guillermoandrae** [comment]
Haven't forgotten about this... just been tied up in other stuff. I'll try to get to it tonight.

**Comment by guillermoandrae** [comment]
My changes are here: https://github.com/guillermoandrae/graph-kit-php/tree/test-suite

@jeremykendall you'll need to merge in the other pull request so I can finish. If you can do it today, I can finish it up tonight.

**Comment by jeremykendall** [comment]
@guillermoandrae Added a comment on #20 RE: merging. Sorry for the delay :-(

**Comment by jeremykendall** [comment]
@ikwattro Rebase against develop?

@guillermoandrae #20 merged.
